Security enhancement - support for Keycloak

ianwallen opened the pull request:
Add security enhancement - Keycloak/OpenID Connect support

To perform a quick test against an existing keycloak setup the following can be done.
- Update  `web/src/main/webapp/WEB-INF/config-security/config-security.xml` to enable keycloak
- build and startup the geonetwork application with https and the following environment variables set.

```
KEYCLOAK_AUTH_SERVER_URL={keycloak url}
KEYCLOAK_REALM={realm name}
KEYCLOAK_RESOURCE={client name}
KEYCLOAK_SECRET={client secret}
KEYCLOAK_DISABLE_TRUST_MANAGER={true|false}
```

If using example setup from the following.
https://www.keycloak.org/getting-started/getting-started-docker

Then the values would be 
```
KEYCLOAK_AUTH_SERVER_URL=http://localhost:8080/auth
KEYCLOAK_REALM=myrealm
KEYCLOAK_RESOURCE=myclient
KEYCLOAK_SECRET={get secret from going to client configuration under Clients -> myclient -> Credentials}
KEYCLOAK_DISABLE_TRUST_MANAGER=true
```
If testing with local host then you may want to set `KEYCLOAK_DISABLE_TRUST_MANAGER=true`  For production it is recommended to use HTTPS and also set this value to false (false is the default if not supplied)

You can also setup more advance keycloak settings by editing the following
`web/src/main/webapp/WEB-INF/config-security/keycloak.json`

**Geonetowk client URL configurations**
Ensure that when you configure your client that you setup the valid redirect uris to your geonetwork installation.
i.e. https://localhost:8443/geonetwork/*
If this is not setup correctly you may get and error indicating that a wrong redirect uri was supplied.
Also if wanting to test the client backchannel logout then ensure that the admin URL is also set to the geonetwork installation.
Here is a sample configuration.

![image](https://user-images.githubusercontent.com/1868233/99267897-e7d90980-27fa-11eb-80cb-5570d1bdda0b.png)

### Sample user/role/group setup
(Assuming client is called myclient)

**Sample Role setup**
In your client role settings (clients -> myclient -> roles). Add the following roles

- Administrator
- RegisteredUser
- Guest
- sample:UserAdmin
- sample:Reviewer
- sample:Editor
- sample:RegisteredUser

**Sample Group configuration**
- Go to keycloak groups (left menu).
- Create a new group called "Administrator"
- Edit the group. Go to Role Mappings -> Client Roles (myclient) -> select the administrator roles and click on "Add selected"
Any user part of the Administrator group will not be an administrator for the myclient applicaton.

**Sample User configuration**

- Go to keycloak users (left menu)
- Add user if not already done so.  Then go to that user.

- Go to role Mappings -> Client Roles (myclient) -> select the available roles to be applied and click on "Add selected"

or 

- Go to Groups -> Available Groups -> Click on the Administrator Group and then click on "Join"



### Single Signout
If testing Single Signout, you may need to ensure that you have proper certificates for keycloak and the geonetwork application.
And ensure the truststore has been setup https://www.keycloak.org/docs/4.8/server_installation/index.html#_truststore
Note: we had some difficulties testing this using localhost due to the https requirements and our corporate network did not allow us to add root certificates required for the truststore.  So we only tested this on a deployed server environment with real domain names.

You will also need to ensure that the Admin URL is configured as identified in the URL Configurations above.

To test Single Signout, you generally require multiple clients on the same realm.  However it is possible to test this by killing the sessions from within Keycloak. So in our case, we initially tested this logic by first login into Geonetwork. - Then go to keycloak Users - and locate the user then go to the Sessions tab.  Then when you logout the user, the system should send a client backchannel request to admin url specified in the url configuration.  This should terminate your geonetwork session. If you then refresh your GN session - you should no longer be logged in.  Note: This will only work if the admin url can be successfully called from keycloak server.  So if there are issues with this functionality, you may want to check the keycloak logs for errors.

### Identity provider (IDP) issues
If testing with and IDP that does not support BackChannel logout you may want to use the parameter .  In this case you may want to add the parameter KEYCLOAK_IDPLOGOUTURL.  It has to be in the format similar to https://idp.example.com/logout?redirect={RedirecUrl}.  {RedirecUrl} is required in the url.
For example. In our case were are using Azure Active Directory (AD) for testing. Azure AD does not support BackChannel logout so we used the following url for KEYCLOAK_IDPLOGOUTURL.
`https://login.microsoftonline.com/common/oauth2/logout?post_logout_redirect_uri={RedirecUrl}`



Comments/testing welcome

** Future changes **
 - Add documentation - will do this after changes have been accepted.
 - Investigate pushing GN group changes back to keycloak  *or* block group changes if only managed via keycloak.
 - Roles are currently assigned in the format of group:role  - i.e. sample:editor - investigate better option

zoran995 commented:
Hi Ian,
This looks really nice. I just have a few questions regarding proper settings:
1. We need to create client roles in Keycloak (administrator, reviewer, editor) to have that role in Geonetwork, or is it something specific for integration?
2. Is it possible to set single logout, so user gets logged out when session is closed from another endpoint (for example GeoServer, or another app integrated with Keycloak)?

I am still trying to figure out group settings (my users, roles and groups are loaded from LDAP/AD to Keycloak).

ianwallen commented:
@zoran995 

1. yes you need to created client roles in keycloak.  For Administrator, registeredUser/Guest these are applied system wide so these can be simple roles.  But if you want to grant a group role then you need to create a keycloak role so that it is in the format of {group}{rolegroupseparator}{role}  -i.e. "sample:editor" which would grant the user editor role on the sample group.  I'm not 100% happy with using formatted role name but I have not found a better way. If someone has a better solution then I would love to hear it

2. The single signout is set in the keycloak software. I'm not sure where the settings are located. I will ask my colleagues to provide a url on where the configuration is located in keycloak.

zoran995 commented:
A quick thought about roles and groups. In keycloak it is possible to define a user group membership mapper that will add an array of groups available for that user, something like this:
```
  "name": "Zoran Kokeza",
  "groups": [
    "/OSGeo",
    "/FAO"
  ]
```
Maybe there could be an option in settings to use this mapper for group membership and that user gets only privileges for those groups. Using this option user with role editor will get editor role in all groups he is member of. This limits the possibilities of geonetwork rules definition where user can have different rules for different groups, so I am not sure how interesting it is to implement this. Something similar is done with LDAP configuration.

ianwallen commented:
@zoran995 
Thank you for the suggestion - I will investigate this option.

Regarding the single signout issue.  You need to enable backchannel logout on the IDP then make sure all the clients have an admin url and you have to hope that the IDP supports backchannel logout.
Search for **backchannel logout**  in the documentation https://www.keycloak.org/docs/latest/securing_apps/. Hope this helps.

zoran995 commented:
I tried to set up backchannel logout but constantly getting HTTP 500 as a response from Geonetwork. Everything is set behind NGINX reverse proxy. I have two error logs that might help in tracking what is going on: [geonetwork_tomcat](https://gist.githubusercontent.com/zoran995/03918fed0fe25a8e11c403020bea745f/raw/b03892a729a4267a4b81e06df342adca2786a482/tomcat_geonetwork_backchannel_error.log) and [keycloak](https://gist.githubusercontent.com/zoran995/8294136e6277d634f872a344ab33efc1/raw/54569af065cca6aa0ed1a35d879e367eef43b883/keycloak_backchannel_geonetwork_error.log)

ianwallen commented:
@zoran995 
What version of keycloak are you testing with?
We have tested with version 4.8.3.
Also what is your client configuration?

Looking at the keycloak code.  It seems that they are not checking the value userSessionManagement to ensure it is not null.  So it is possible that this value was null and that is what caused the NPE error.

https://github.com/keycloak/keycloak/blob/master/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/PreAuthActionsHandler.java#L147

Why is it null - I do not know?

We had some issues with the logout so we extended the existing keycloaklogout.
If you are using a newer version of keycloak they may have fixed the issues.  So you may want to try changing the bean reference
https://github.com/ianwallen/core-geonetwork/blob/add_keycloak_support/web/src/main/webapp/WEB-INF/config-security/config-security-keycloak.xml#L83

from `org.fao.geonet.kernel.security.keycloak.KeycloakLogoutHandler` to `org.keycloak.adapters.springsecurity.authentication.KeycloakLogoutHandler`

zoran995 commented:
We are using the Keycloak v10.0.2, but just tried with 4.8.3 and still getting the same NPE 😢. The client is  set like the following: 
- protocol: openid-connect, 
- access  type: confidential. 
- Standard Flow: enabled, 
- implicit flow: disabled,
- Direct Access Grants: enabled,
- Service Accounts Enabled: disabled,
- Authorization Enabled: disabled

There is a https://keycloak.discourse.group/t/spring-security-logout-from-keycloak-not-propagated/3040, where a similar issue is described, but more investigation will be needed to see how to apply the solution.

zoran995 commented:
@ianwallen I tried adding to `HttpSessionManager`, but tracking it down it doesn't know of the user session. The standard way to get a user session would be to add
```
<listener>
    <listener-class>org.fao.geonet.kernel.security.keycloak.HttpSessionEventPublisher</listener-class>
 </listener>
```
into a `web.xml`, as described in keycloak docs for spring-security, but it throws an error. This part is pretty complex in geonetwork, is there another way to get list of user session so it can get invalidated.

Relating the KeycloakLogoutHandler it's called only when I click on the logout button in geonetwork, and it works properly to logout user from keycloak.

ianwallen commented:
@zoran995 I got the Single signout working.  If you try the latest changes you can give it a try.

zoran995 commented:
Just done some testings and single signout works perfectly. Using the default settings a user will be required to login to geonetwork to access the content? I tried changing the `loginType` to link but it returns me to keycloak defaults login type as a form, instead of link type where the user should be redirected to keycloak authentication page. You probably forgot to remove references to `isSSO` in [those lines](https://github.com/geonetwork/core-geonetwork/pull/4931/files#diff-c0a35cff5778e2042a7d6a474fd245bec2253a4b627efab8a2502102c7a25a05R200-R213). Will do more testing during in the following weeks but the rest seems to work flawlessly.

ianwallen commented:
@zoran995 - Thank you very much for testing.
I had accidently forgotten to commit the isSSO changes to that file. Just committed the changes now.

jodygarnett commented:
Update from chat:

* Thanks for the review @zoran995 - that is great!
* Jose to review PR to check for any interactions with rest of the codebase, and then merge :)

Thanks for this functionality Ian, it is a great addition (and give a nice way for geoserver and geonetwork to do single sign on)

josegar74 commented on core/src/main/java/org/fao/geonet/kernel/security/SecurityProviderConfiguration.java:
```diff
@@ -0,0 +1,99 @@
+ */
+public interface SecurityProviderConfiguration {
+
+	public enum LoginType {
```
Does this configuration affects `LDAP` or `CAS` providers?

ianwallen replied:
It was mostly used for Shibboleth for the SSO parameters.  For all cases it would check for the existence of the bean and if the bean was not found then then it continued processing as it did before.  So unless LDAP or CAS was using Shibboleth configuration options, they should not be affected by these changes.

Once this PR is approved then it would probably be a good idea if LDAP or CAS also used this class if there is a need to support common configurations for SSO.

josegar74 commented:
@ianwallen can you add some documentation to describe the configuration?

I want to test with https://www.keycloak.org/getting-started/getting-started-docker

But from some of the values required:

```
KEYCLOAK_AUTH_SERVER_URL=https://localhost/auth
KEYCLOAK_SECRET={client secret}
KEYCLOAK_RESOURCE={client name}
KEYCLOAK_REALM={realm name}
```

`KEYCLOAK_SECRET` or `KEYCLOAK_RESOURCE` are not very clear what should contain. I understand that probably for users familiar with KeyCloak can be quite obvious, but if can be updated. 

Also with some examples on users / roles / groups configuration, would help to test the feature.

ianwallen commented on core/src/main/java/org/fao/geonet/kernel/security/keycloak/keycloakPreAuthActionsLoginFilter.java:
```diff
@@ -0,0 +1,62 @@
+package org.fao.geonet.kernel.security.keycloak;
```
header added

ianwallen commented on core/src/main/java/org/fao/geonet/kernel/security/keycloak/KeycloakHttpSessionManager.java:
```diff
@@ -0,0 +1,85 @@
+
+// Changes were taken from org.keycloak/keycloak-adapter-core/11.0.1 as there were bugs in current version
+// that we prevending the Admin URL client backchannel logout from working.
+// If upgrading maven dependency then it may be possible that the most recent KeycloakHttpSessionManager can be used
```
I updated the comments to make it more clear as to why this was required.
It is mostly because the WebApplicationContextUtils.getWebApplicationContext(servletContext) returns null and it was causing NPE.  It seems like the the application is not registered as a WebApplicationContext so I had to modify the code to use ApplicationContextHolder.get() and this fixed the issue.

Maybe in the future if the application is registered correctly as a WebApplicationContext then this file can be removed.

Is this correction OK - or should there be a fix to register the application correctly as a WebApplicationContext?

ianwallen commented on core/src/main/java/org/fao/geonet/kernel/security/listener/HttpSessionEventPublisher.java:
```diff
@@ -0,0 +1,87 @@
+import javax.servlet.http.HttpSessionEvent;
+import javax.servlet.http.HttpSessionListener;
+
+/* This is just a copy of the org.springframework.security.web.session.HttpSessionEventPublisher
```
I updated the documentation to provide more details as to why this was required. It also references the bug that describes the issue.

This is required because our code was based on 4.8.3 which is the version that RedHat SSO 7.3 uses. This is our target software at the moment.  It was noticed that RedHat SSO has released a newer version that now uses the newer drivers. 
Our intention is to upgrade to the newer version of the software in the new year.  At that time I will create a PR to upgrade to newer version of the keycloak adapter and I can will be able to remove this file.

ianwallen commented on core/src/main/java/org/fao/geonet/kernel/security/keycloak/KeycloakUtil.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public static String getSigninPath() {
+        if (signinPath == null) {
```
It is used for caching.

`public static String signinPath = null;
`

It checks to see if signinPath still contains the initial value and if so then sets the values based on configuration values from [config-security.properties](https://github.com/geonetwork/core-geonetwork/blob/master/web/src/main/webapp/WEB-INF/config-security/config-security.properties#L27).
`loginForm=/signin?node=@@nodeId@@`

Subsequent calls should just return the value.

The intent was to use existing configuration values instead of hardcoding the value to "/signin"  But I did noticed that there are some hard coded values of "/signin" elsewhere in the code(mostly in Angular) so I'm not sure how easy it would be to change this property anyway

ianwallen commented:
@josegar74 I updated the comments on this PR to better explain how to set the configuration.  And I also added some sample user /group/role configuration.

zoran995 commented:
> Tested with the default database authentication and seem no unwanted side effect, with KeyCloak enabled I get redirected to KeyCloak, but get the following error in KeyCloak:
> 
> ```
> We are sorry...
> Invalid parameter: redirect_uri
> ```

@josegar74 KeyCloak is very security sensitive to the URL connection is coming from (which is classified as `redirect_uri` and the user will be redirected to it on successful login). Have you set the correct `Valid Redirect URIs` for KeyCloak client created for geonetwork (i.e. http://localhost:8080/geonetwork/*). Also, the protocol for the redirect_uri must be correct.

ianwallen commented:
Thank you @zoran995 , I just updated the PR description to provide the url settings to be placed in keycloak. I also added some other instructions related to keycloak setup.  It would be appreciated if you could review the PR descriptions to see if missed anything that could cause issues for the reviewers.

I find the client backchannel logout/IDP backchannel logout caused us the most issues. Mostly because I find the documentation is not clear on identifying the differences between both of these backchannel logout.  Also, we had issues with setting up the proper security keystores for this to work in our environment.

zoran995 commented:
@ianwallen thank you very much for all of this, you pretty much cover everything. I am just little wondering if autologin should be default login type because it will redirect the user to the KeyCloak login page and he can't access the geonetwork page without logging. We changed the configuration to use `LINK` type.

I thought that IDP backchannel logout is out the scope of this and that KeyCloak handles it but now I see that you included it, so I probably had the wrong assumption and it is really nice to see that also supported. It can be really tricky to clearly identify them in the documentation and articles (IDP is used if someone already has some SAML or OIDC service and needs to enable login of those users through Keycloak). 
- Regarding the SSL and keystores - We are using NGINX in front of our application and configure SSL on it, and according to the [KeyCloak documentation](https://www.keycloak.org/docs/latest/server_installation/#_setting_up_ssl) it is enough to just set it there, this will save most of the time for certificates and headaches later when updating. When using the approach behind a reverse proxy it is necessary to pay attention to https://www.keycloak.org/docs/latest/server_installation/#enable-https-ssl-with-a-reverse-proxy, as some KeyCloak internal URLs might not work properly (not particularly important for review).
- some notes regarding the configuration
  - confidential client is recommended type of client to be created on KeyCloak 
  - `keycloak.json` values can be found in the installation tab of client config on KeyCloak 
![image](https://user-images.githubusercontent.com/19208948/99280928-35bd3400-2832-11eb-8efb-2ed56491a44d.png)
  - confidential port is important when running behind a reverse proxy for most use cases as you need to set it to propper values for all redirects to work (for example 443) - not sure for other use cases. 

To anyone wondering about version conflict of KeyCloak, we are using this with KeyCloak 11.0.2

zoran995 commented:
> and give a nice way for geoserver and geonetwork to do single sign on

Yeah single sign-on works without any issue using keycloak, only Geoserver integration could use little more work to support single sign out (something similar Ian did here) but I don't think that there are many users use the session with OGC services (this is just a side note if somebody reads this PR comments when implementing SSO).

ianwallen commented:
@zoran995
> I am just little wondering if autologin should be default login type because it will redirect the user to the KeyCloak login page and he can't access the geonetwork page without logging. We changed the configuration to use `LINK` type.

When I first started to add the code, we were using LINK option.  But then when management/clients saw the changes, they asked that guest not have access to the system as they want to ensure that only the organization personal have access.  So I created a new type for autologin and did not think about changing it back. 
As I think about it, you may be correct as it seems to be more inline with how geonetwork currently works. 
I will push a change to make  `LINK` the default.

Thank you for your input

josegar74 commented:
Thanks for the documentation updates to setup KeyCloak to test the pull request. Once the change to make `LINK` the default, it can be merged.

ianwallen commented:
@josegar74 The default login type was changed to Link

ianwallen commented:
@josegar74 
Could you allocate a target milestone for this PR. I'm assuming 3.10.6 or 3.12.

pvgenuchten commented:
used text in this issue to create https://github.com/geonetwork/doc/pull/175 (hope i don't duplicate work)
